### ISSUE TITLE
Core refactor: integrated flows as states instead of storages

### ISSUE THREAD

**Issue by SouthEndMusic** [Deltares/Ribasim#1808]
# What

I propose, as discussed earlier with @visr and @Huite, to reformulate the ODE problem that is solved by the core somewhat. It must be emphasized that this does not fundamentally change the problem, in the sense that exact solutions to both formulations are the same. Modelers should not see large changes in output before and after this refactor.

Currently the states that the core solves for over time are the storages in the basins (and the error integral for `PidControl`, but this will likely remain unchanged and thus wil not be mentioned again in this issue). The proposal is to have the integrated flow over each edge from the start of the simulation to the 'current time' as the state:

$$
u_i = \int_{t_0}^t Q_i(\mathbf{u},p,t')\text{d}t'
$$

where the index $i$ denotes an edge (or a basin for the basin forcings). This leads to a very simple set of differential equations

$$
\frac{\text{d}u_i}{\text{d}t} = Q_i(\mathbf{u},p,t),
$$

where the $Q_i$ are given by the formulas in the reference docs for the nodes.

# How

As far as I can oversee now before diving into this, the refactor will not be hugely complex or intrusive. Roughly speaking, currently our right hand side function `water_balance!` takes as argument the storages, computes from that the levels, formulates from that the flows, and formulates from that the `du` (i.e. the time derivative of the storages).

In the new formulation, `water_balance!` takes in the integrated flows, computes from that in combination with the initial storages the current storages, computes from that the levels, and formulates from that the flows which go directly into `du` (which in the new formulation denotes the time derivative of the integrated flows, i.e. the flows themselves).

# Why

Generally speaking, we care a lot about integrating and averaging flows accurately (for e.g. computing the water balance error within Ribasim, a closing water balance when coupling, inputs for the allocation algorithm). We have however a rather crude way of computing these, which shadows what the ODE solver is doing. So, why not let the solver 'care' about integrating individual flows more explicitly?

### Why it is now bad

Over the past few days I have discussed a lot with @visr and @Huite what the water balance error should precisely mean and how it can be implemented into the solver more deeply. I did this because I was working on this issue: https://github.com/Deltares/Ribasim/issues/1413. The current formulation of the water balance error compares 2 things:
1. The total flow related to a basin as expected from its storage change over a timestep;
2. The total average flow related to a basin as computed from formulated flows given the storages (and so levels) interpolated over that timestep.

The integration for point 2 is either low in accuracy or adds expensive computation, and it is also not so clear how the storage should be interpolated over the timestep. This makes that I don't have a lot of confidence in this water balance error formulation. I also implemented standard tolerances on this (absolute error 1 L/s, relative error 1e-3) which from my current experience I am sure realistic  models are not going to pass. 

### Why the new formulation is better

In the new formulation the old water balance as computed above is satisfied by definition. It's also easier to relate the solver error to the schematization because each error is for the flow as formulated by one node, instead of possibly many nodes connected to a basin.

As I understand it, the goal of the water balance error is to quantify how much water is erroneously 'created' or 'destroyed' in the model, preferably on a 'local' basis (i.e. in the current formulation on a per basin basis). In the proposed formulation, each timestep dictates how much water moves between basins or in and out of the model from which the storage changes are then explicitly computed, so all storage change is accounted for. The only error that is left is how well the computed flows match the implemented physics, given by the solver error.

## Exact integration

One thing that bugged me when working on flow integration is that for some edges (i.e. connected to `FlowBoundary`) and incoming basins forcings (i.e. precipitation, drainage) the integrated flow can be computed exactly, but we don't do that yet. In the new formulation we can simply leave those states out and implement the exact integration.

# Performance and accuracy

- As pointed out by @Huite, this formulation has more states, which might negatively impact performance. However, it's hard to oversee all performance impacts so this should just be tested. To reduce the number of states, the in- and outflow of conservative nodes should be represented by a single state.
- As discussed before in the context of exposing integrated flows via BMI: integration over long periods of time can lead to loss in accuracy because new contributions become small with respect to the total integrated flow. This can be solved in the known way, resetting the state to $0$ and transferring the cumulated flow to the 'initial condition'.

**Comment by SouthEndMusic** [comment]
We do have to rework `isoutofdomain` and related features a bit, because of the computation form integrated flows to storages.

**Comment by SouthEndMusic** [comment]
TL;DR of this refactor proposal:
- The water balance will be closed by definition
- Flow integration, needed for e.g. coupling and allocation, will be accurate
- It's easier to relate the solver error to a single node rather than all nodes connected to a basin
- Performance impacts remain to be seen

**Comment by SouthEndMusic** [comment]
Some more in-depth on the how:

## Data structures

- **The graph metadata**: `flow`, `flow_prev`, `flow_integrated` and `saveat` can be removed (that last one seems to be unused already)
- **The state vector**: the `storage` component can be removed and a `flow` component must be added. As a nice feature this can also be broken up into a component for each conservative node and separate in- and outflow components for non conservative nodes. Also components for outgoing forcings.
- **The `Basin` struct**: all `vertical_flux_*` can be removed apart from `vertical_flux_from_input`. An `inial_storage` vector and a `current_storage` `cache` must be added

## The right hand side (`water_balance!`)

- In `set_current_basin_properties!` compute the current storage from `u.flow_*`, `basin.inital_storage` and exact integrated flows. Note that the exact integrated flows only depend on `t`, so it would be nice if we could find a way to only compute these once for a proposed `dt`
- In `formulate_basins!` set the forcings directly into `du`. To be decided: does each forcing type get its own state?
- In all `formulate_flow!` methods set the flow directly into the corresponding `du.flow_*`
- `formulate_du!` and `formulate_du_pid_controlled!` can be removed. Some care must be taken with what must be kept in terms of formulating flows in a particular order because some flows depend on other flows (`PidControl`, `ContinuousControl`)

## Callbacks

- `integrating_flows_cb` can be removed (🙏). Average flows used for allocation input can be computed on demand given the `integrator.sol`. 
- Not sure why `update_vertical_flux!` is called in `basin_cb` (it is not documented), as that is also part of `water_balance!`. I'm quite sure this was only done for flow integration so it can be removed. The same holds for `set_previous_flows!`.

## Output

Storages, levels, average flows, average forcings and subgrid levels can (still) be saved with saving callbacks. Note that storages and levels depend on the last call of `water_balance!` so to be sure to save accurate values they should be calculated again using the new value of `u` before saving. That also helps with proper values of the storages and levels being read by a coupled model after a Ribasim timestep.

## BMI

We currently expose integrated basin forcings and user demand inflows via the BMI. For this we can probably just expose the relevant components of `u`. If we decide to indeed reset the state once in a while (tbd) as mentioned in the original post above, this has to be aligned with what a coupled model expects here.
